Sources/NIOExtras/HTTP1ProxyConnectHandler.swift

```diff
@@ -0,0 +1,299 @@
+        switch self.state {
+        case .headReceived(let timeout):
+            timeout.cancel()
+            self.state = .completed
+            self.promise.succeed(())
+
+        case .failed:
+            // ran into an error before... ignore this one
+            break
+        case .initialized, .connectSent, .completed:
+            preconditionFailure("Invalid state: \(self.state)")
+        }
+    }
+
+    private func failWithError(_ error: Error, context: ChannelHandlerContext, closeConnection: Bool = true) {
+        self.state = .failed(error)
+        self.promise.fail(error)
+        context.fireErrorCaught(error)
+        if closeConnection {
+            context.close(mode: .all, promise: nil)
+        }
+    }
+
+    /// Error types for ``HTTP1ProxyConnectHandler``
+    public struct Error: Swift.Error, Equatable {
+        fileprivate enum Storage: Equatable, Hashable {
+            case proxyAuthenticationRequired
+            indirect case invalidProxyResponseHead(head: HTTPResponseHead)
+            case invalidProxyResponse
+            case remoteConnectionClosed
+            case httpProxyHandshakeTimeout
+            case noResult
+
+            // compare only the kind of error, not the associated response head
+            @inlinable
+            static func == (lhs: Self, rhs: Self) -> Bool {
+                Kind(lhs) == Kind(rhs)
+            }
+
+            @inlinable
+            public func hash(into hasher: inout Hasher) {
+                hasher.combine(Kind(self))
+            }
+        }
+
+        fileprivate enum Kind: Equatable, Hashable {
+            case proxyAuthenticationRequired
+            case invalidProxyResponseHead
+            case invalidProxyResponse
+            case remoteConnectionClosed
+            case httpProxyHandshakeTimeout
+            case noResult
+
+            init(_ storage: Storage) {
+                switch storage {
+                case .proxyAuthenticationRequired:
+                    self = .proxyAuthenticationRequired
+                case .invalidProxyResponseHead:
+                    self = .invalidProxyResponseHead
+                case .invalidProxyResponse:
+                    self = .invalidProxyResponse
+                case .remoteConnectionClosed:
+                    self = .remoteConnectionClosed
+                case .httpProxyHandshakeTimeout:
+                    self = .httpProxyHandshakeTimeout
+                case .noResult:
+                    self = .noResult
+                }
+            }
+        }
+
+        final class Location: Sendable {
+            public let file: String
+            public let line: UInt
+            init(file: String, line: UInt) {
+                self.file = file
+                self.line = line
+            }
+        }
+
+        fileprivate let error: Storage
+        let location: Location
+
+        fileprivate init(error: Storage, file: String = #file, line: UInt = #line) {
```

Comment:
Unfortunately we need all the `static let`s to be `static func`s below which also take a `file` and `line` so that we capture the `file` and `line` where the error was created (otherwise it will just be the line of the `static let`).

I recommend removing the defaults here: to enforce that we pass through through to the init.